File: app/src/androidTest/java/com/tillchen/fyberchallenge/UITest.java
package com.tillchen.fyberchallenge;

import android.content.Context;
import android.os.SystemClock;

import androidx.test.espresso.intent.Intents;
import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.ext.junit.rules.ActivityScenarioRule;
import androidx.test.filters.LargeTest;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.tillchen.fyberchallenge.activities.MainActivity;
import com.tillchen.fyberchallenge.activities.OfferListActivity;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.clearText;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.hasErrorText;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static androidx.test.espresso.intent.Intents.intended;
import static androidx.test.espresso.intent.matcher.IntentMatchers.hasComponent;

@RunWith(AndroidJUnit4.class)
@LargeTest
public class UITest {

    @Rule
    public ActivityScenarioRule<MainActivity> activityRule
            = new ActivityScenarioRule<>(MainActivity.class);

    private String getResourceString(int id) {
        Context targetContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        return targetContext.getResources().getString(id);
    }

    @Test
    public void testTextFieldsArePrePopulated() {
        onView(withId(R.id.editTextAppId)).check(matches(withText(R.string.test_app_id)));
        onView(withId(R.id.editTextUserId)).check(matches(withText(R.string.test_user_id)));
        onView(withId(R.id.editTextSecurityToken))
                .check(matches(withText(R.string.test_security_token)));
    }

    @Test
    public void testEmptyAppId() {
        onView(withId(R.id.editTextAppId)).perform(clearText());
        onView(withId(R.id.buttonShowOffers)).perform(click());
        onView(withId(R.id.editTextAppId))
                .check(matches(hasErrorText(getResourceString(R.string.empty_app_id))));
    }

    @Test
    public void testEmptyUserId() {
        onView(withId(R.id.editTextUserId)).perform(clearText());
        onView(withId(R.id.buttonShowOffers)).perform(click());
        onView(withId(R.id.editTextUserId))
                .check(matches(hasErrorText(getResourceString(R.string.empty_user_id))));
    }

    @Test
    public void testEmptySecurityToken() {
        onView(withId(R.id.editTextSecurityToken)).perform(clearText());
        onView(withId(R.id.buttonShowOffers)).perform(click());
        onView(withId(R.id.editTextSecurityToken))
                .check(matches(hasErrorText(getResourceString(R.string.empty_security_token))));
    }

    @Test
    public void testShowOffers() {
        Intents.init();
        onView(withId(R.id.buttonShowOffers)).perform(click());
        // We can improve this with: https://developer.android.com/training/testing/espresso/idling-resource
        SystemClock.sleep(5000);
        intended(hasComponent(OfferListActivity.class.getName()));
    }

}

File: fyberchallengesample/src/androidTest/java/com/tillchen/fyberchallengesample/UITest.kt
package com.tillchen.fyberchallengesample

import android.os.SystemClock
import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.intent.Intents
import androidx.test.espresso.intent.matcher.IntentMatchers
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.ext.junit.rules.ActivityScenarioRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.tillchen.fyberchallenge.activities.OfferListActivity
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Instrumented test, which will execute on an Android device.
 *
 * See [testing documentation](http://d.android.com/tools/testing).
 */
@RunWith(AndroidJUnit4::class)
class UITest {

    @get:Rule
    val activityRule = ActivityScenarioRule(MainActivity::class.java)
    private fun getResourceString(id: Int): String {
        val targetContext = InstrumentationRegistry.getInstrumentation().targetContext
        return targetContext.resources.getString(id)
    }

    @Test
    fun testTextFieldsArePrePopulated() {
        Espresso.onView(withId(R.id.editTextAppId)).check(ViewAssertions.matches(ViewMatchers.withText(R.string.test_app_id)))
        Espresso.onView(withId(R.id.editTextUserId)).check(ViewAssertions.matches(ViewMatchers.withText(R.string.test_user_id)))
        Espresso.onView(withId(R.id.editTextSecurityToken))
            .check(ViewAssertions.matches(ViewMatchers.withText(R.string.test_security_token)))
    }

    @Test
    fun testEmptyAppId() {
        Espresso.onView(withId(R.id.editTextAppId)).perform(ViewActions.clearText())
        Espresso.onView(withId(R.id.buttonShowOffers)).perform(ViewActions.click())
        Espresso.onView(withId(R.id.editTextAppId))
            .check(ViewAssertions.matches(ViewMatchers.hasErrorText(getResourceString(R.string.empty_app_id))))
    }

    @Test
    fun testEmptyUserId() {
        Espresso.onView(withId(R.id.editTextUserId)).perform(ViewActions.clearText())
        Espresso.onView(withId(R.id.buttonShowOffers)).perform(ViewActions.click())
        Espresso.onView(withId(R.id.editTextUserId))
            .check(ViewAssertions.matches(ViewMatchers.hasErrorText(getResourceString(R.string.empty_user_id))))
    }

    @Test
    fun testEmptySecurityToken() {
        Espresso.onView(withId(R.id.editTextSecurityToken)).perform(ViewActions.clearText())
        Espresso.onView(withId(R.id.buttonShowOffers)).perform(ViewActions.click())
        Espresso.onView(withId(R.id.editTextSecurityToken))
            .check(ViewAssertions.matches(ViewMatchers.hasErrorText(getResourceString(R.string.empty_security_token))))
    }

    @Test
    fun testShowOffers() {
        Intents.init()
        Espresso.onView(withId(R.id.buttonShowOffers)).perform(ViewActions.click())
        // We can improve this with: https://developer.android.com/training/testing/espresso/idling-resource
        SystemClock.sleep(5000)
        Intents.intended(IntentMatchers.hasComponent(OfferListActivity::class.java.name))
    }
}
